KCORE-414; Make request/response error handling more conventional

hachikuji opened the pull request:
The purpose of this patch is to get rid of the clumsy error handling in `maybeHandleError` and establish more consistent conventions for request/response handling.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -337,26 +342,39 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
-            logger.debug("Ignoring vote response {} since we already became leader for epoch {}",
+    private Optional<Errors> handleVoteResponse(
+        RaftResponse.Inbound responseMetadata
+    ) throws  IOException {
```
nit: extra space

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -765,18 +840,16 @@ private void handleInboundMessage(RaftMessage message, long currentTimeMs) throw
                                          int destinationId,
                                          Supplier<ApiMessage> requestData) throws IOException {
-        ConnectionCache.ConnectionState connection = connections.getOrCreate(destinationId);
+        ConnectionState connection = connections.getOrCreate(destinationId);
```
nit: could be final

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -337,26 +342,39 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
+                    maybeBecomeLeader(state);
+                } else {
+                    state.recordRejectedVote(remoteNodeId);
+                    if (state.isVoteRejected()) {
```
Thinking about this a bit more after #309 , it's a bit different between "adding the jitter to the election timeout" v.s. "backoff jitter time before issuing a new election" since in the first case (our current impl) we could start sending the vote request right after, where as in the latter case, we first wait for jitter, then start sending the vote request.

So back to this example, I think we should reset the electionTimer to `jitter` so that we can become a candidate after jitter time instead of letting the whole election time to be elapsed. And in `becomeCandidate` we should not use `election + jitter` but should always use `election` (I think in the original paper we do not need a random jitter for election itself, just the jitter for backing off?)

hachikuji replied:
Let's discuss this separately. I think my assumption has been that jitter is small relative to the election timeout. If that is not the conventional approach, then we should change it.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -409,6 +446,22 @@ private EndQuorumEpochResponseData handleEndQuorumEpochRequest(EndQuorumEpochReq
+
+        if (handleNonMatchingResponseLeaderAndEpoch(response.leaderEpoch(), optionalLeaderId(response.leaderId()))) {
+            return Optional.of(error);
+        } else if (error == Errors.NONE) {
```
For both Begin / EndQuorum request, we can also return FENCED_LEADER_EPOCH and INVALID_REQUEST which should be handled before `handleUnexpectedError` right? I think we should not just backoff on connection cache for these two?

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -740,22 +804,33 @@ private void handleResponse(RaftResponse.Inbound response) throws IOException {
 
     private void handleRequest(RaftRequest.Inbound request) throws IOException {
-        ApiMessage requestData = request.data();
+        ApiKeys apiKey = ApiKeys.forId(request.data.apiKey());
```
This is for the old line 726 `validateLeaderOnlyRequest`: it should be `validateLeaderOnlyResponse`.

hachikuji replied:
I think one of the methods should be `validateVoterOnlyRequest` and the other should be `validateLeaderOnlyRequest`. I made this change. Was there another invalid naming?

guozhangwang replied:
That's good enough.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -686,21 +730,41 @@ private boolean handleNonMatchingResponseLeaderAndEpoch(int epoch, OptionalInt l
+
+        ConnectionState connection = connections.getOrCreate(response.sourceId());
+        if (errorOpt.isPresent()) {
+            connection.onResponseError(response.correlationId, currentTimeMs);
```
nit: this makes me thinking, before calling `onResponseError` could we log a DEBUG indicating which error cause us to backoff on that connection?

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -686,21 +730,41 @@ private boolean handleNonMatchingResponseLeaderAndEpoch(int epoch, OptionalInt l
+        }
+
+        ConnectionState connection = connections.getOrCreate(response.sourceId());
+        if (errorOpt.isPresent()) {
```
If we check `errorOpt.isPresent() && errorOpt.get() != NONE` here, then we would not need the special handling in fetch quorum records of `handleNonMatchingResponseLeaderAndEpoch` right?

hachikuji replied:
I was using the `Optional` type here to represent the presence of absence of an error, so the need to check for `NONE` would defeat that. Let me think a little about this. Perhaps we should just return `Errors` directly.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -564,11 +640,31 @@ private FindQuorumResponseData handleFindQuorumRequest(FindQuorumRequestData req
+            // Unlike other API types, FindQuorum may indicate have a higher epoch or
+            // a different leaderId than our own and still return an error code of NONE,
+            // so we have to check for it
+            if (error == Errors.NONE)
```
See my other comment: logically I think returning an `Optional.empty()` should have the same semantics as returning an `Optional.of(NONE)`.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -765,18 +840,16 @@ private void handleInboundMessage(RaftMessage message, long currentTimeMs) throw
+            handleResponse(response, currentTimeMs);
         } else {
-            throw new IllegalStateException("Unexpected message " + message);
+            throw new IllegalArgumentException("Unexpected message " + message);
```
Why change it to illegal-argument? I think illegal-state is more accurate :P

hachikuji replied:
The difference between these two exceptions can get confusing in my mind. Here I see this as basically a problem of an invalid input, so I thought `IllegalArgumentException` was a better fit.

hachikuji commented:
@guozhangwang @abbccdda Thanks for the comments. I've pushed a couple commits which I believe addresses your comments. In general, we try to be more rigorous with the error handling by identifying all of the possible errors and ensuring they are handled. The core changes are the following:

1) I've added a new INCONSISTENT_VOTER_SET as a more useful error for voter-only requests
2) I've removed the logic to return some errors which should not be possible. For example, there should never be a reason why we return `NOT_LEADER_FOR_PARTITION` from `FetchQuorumRecords` in normal circumstances. A stale fetch request would get fenced by the epoch. And unlike with topic partitions, a follower cannot learn about the leader before the leader itself.
3) I've added some additional validations. For example, we now add checks to make sure that we reject requests which indicate we are the leader when we know we are not. We also validate all offsets and epochs strictly in both FetchQuorumRecords and Vote requests.

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
         } else {
-            if (!quorum.leaderId().isPresent())
-                return false;
             if (!leaderId.isPresent())
```
Could we merge the cases into one boolean expression?
```
return epoch != quorum.epoch() || !leaderId.isPresent() || !quorum.leaderId().isPresent() || leaderId.equals(quorum.leaderId());
```

hachikuji replied:
We could, but would that be easier to follow? It took me a few iterations to get this check right. At least let me collapse the branches here.

hachikuji replied:
Decided to do this after all. Since it is just chaining `||`, I think the suggestion is still clear enough.

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
+            // after backing off.
+            throw new IllegalStateException("Received error indicating inconsistent voter sets");
+        } else if (error == Errors.INVALID_REQUEST) {
+            throw new IllegalStateException("Received unexpected invalid request error");
```
Should we pass the error here as well?

hachikuji replied:
Not sure what the benefit would be. The message already indicates it was an invalid request error.

abbccdda replied:
Oh, you are right.

abbccdda commented on raft/src/test/java/org/apache/kafka/raft/QuorumStateTest.java:
```diff
@@ -151,6 +151,17 @@ public void testCannotBecomeLeaderIfAlreadyLeader() throws IOException {
 
+    @Test
+    public void testCannotBecomeFollowerOfSelf() throws IOException {
+        int otherNodeId = localId + 1;
```
I think `otherNodeId` is not necessary.

hachikuji replied:
Fair enough.

abbccdda commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java:
```diff
@@ -1059,14 +1151,14 @@ public void testFetchShouldBeTreatedAsLeaderEndorsement() throws Exception {
         assertBeginQuorumEpochRequest(epoch);
 
-        FetchQuorumRecordsRequestData fetchRequest = fetchQuorumRecordsRequest(epoch, otherNodeId, 0L);
+        FetchQuorumRecordsRequestData fetchRequest = fetchQuorumRecordsRequest(epoch, otherNodeId, 0L, 0);
```
We don't have to address this in the current PR, but I feel a lot of request/response data generation code paths are overlapping with the core KafkaRaftClient library. Potentially we could consolidate them by putting the data initialization logic into their corresponding request/response classes, like moving this call towards `FetchQuorumRecordsRequest.java`

hachikuji replied:
Yeah, I have the same feeling. I was trying hard not to use objects like `FetchQuorumRecordsRequest` in this implementation because ultimately we will need to remove them. However, frankly speaking, the use of builders all over the place makes for very verbose code. I was thinking instead of moving these helpers into `RaftUtil`.

abbccdda commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java:
```diff
@@ -695,6 +682,111 @@ public void testObserverFindQuorumAfterElectionTimeout() throws IOException {
     }
 
+    @Test
+    public void testInvalidFetchRequest() throws Exception {
```
Nice added coverage! Do we have coverage for the LogTruncationException logic as well? I ran the coverage tool but didn't it gets triggered.

hachikuji replied:
This is a gap I am hoping we can address when we add the `truncate` API to `ReplicatedStateMachine`.

abbccdda replied:
Sg, will get a JIRA.

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
+    private Optional<Errors> validateVoterOnlyRequest(int remoteNodeId, int requestEpoch) throws IOException {
+        if (requestEpoch < quorum.epoch()) {
+            return Optional.of(Errors.FENCED_LEADER_EPOCH);
+        } else if (quorum.isObserver() || !quorum.isVoter(remoteNodeId)) {
```
I feel these two scenarios are not quite the same, as the former indicates a state error of the receiver, while the later indicates a sender's error. Does `UNKNOWN_VOTER` fit better for the latter case?

hachikuji replied:
Originally I did make this into separate errors, but I talked myself out of it because I could not think of a reason why the behavior would be different when the response is received. Either way, I think we're just going to handle this error by backing off and retrying.

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
         }
 
-        // TODO: seems like a weird place to do this...
+        // The request is from a valid voter and indicates a larger epoch,
```
This case combined with L881 looks weird, as if we don't know the identity of the sender, we shall early terminate; but in the meantime if it has a higher epoch, we shall just follow it. It might worth discussing what's the general strategy here, as even through static config, the quorum voters could change so the current receiver may hold stale membership data which prevents it from continuing to serve in the new quorum.

hachikuji replied:
I'll add a comment. As far as I understand it, the expected behavior in the reassignment protocol is to always trust the latest voter state we have seen in the log. But if we see a larger epoch from one of the expected voters, we need to take it.

abbccdda replied:
So the agreement is that sender will retry until it detects the latest membership info right?

hachikuji replied:
Yes, or until the receiver gets the latest membership if the sender is right.

hachikuji replied:
I was thinking about this a little more and it does seem we'll need to come back to it when we implement reassignment. The current validation makes sense only if you assume static voters. The problem is that a node which has been removed from the voter set needs to have some way to find the new voter set. It might not be able to do that if it only trusts its latest voter information. This might suggest the need for a new state, but not sure.

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
-        } else if (hasInconsistentLeader(epoch, leaderId)) {
-            throw new IllegalStateException("Received response with leader " + leaderId +
-                    " which is inconsistent with current leader " + quorum.leaderId());
+        } else if (leaderId.isPresent() && !quorum.leaderId().isPresent()) {
```
Since we now have 3 sub-states for a follower:
```
1. voted
2. fetching
3. unattached
```
Not proposing adding another state enum, should we just define 3 separate `is***Follower` check on `QuorumState` which logic is easier to be reused?

hachikuji replied:
I added a `hasLeader` method to make this a little more concise.

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +746,236 @@ private void updateVoterConnections(FindQuorumResponseData response) {
+     * @param error Error from the received response
+     * @param leaderId Optional leaderId from the response
+     * @param epoch Epoch received from the response
+     * @return Optional value indicating whether the error was handled here. True indicates
```
I feel by listing:
```
Optional.empty
Optional(true)
Optional(false)
```
side by side with each case definition could achieve better clarity. So far for me I need to translate the optional first in mind and then understand the logic.

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -306,10 +314,22 @@ private VoteResponseData buildVoteResponse(Errors error, boolean voteGranted) {
 
-    private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOException {
-        Optional<Exception> errorOpt = validateLeaderOnlyRequest(request.candidateId(), request.candidateEpoch());
+    /**
```
It's a little annoying to duplicate it since that it's hard to keep them in sync, so I just decided to document the expected errors on the request handler.

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -306,10 +314,22 @@ private VoteResponseData buildVoteResponse(Errors error, boolean voteGranted) {
+     * - {@link Errors#FENCED_LEADER_EPOCH} if the epoch is smaller than this node's epoch
+     * - {@link Errors#INCONSISTENT_VOTER_SET} if the request suggests inconsistent voter membership (e.g.
+     *      if this node or the sender is not one of the current known voters)
+     * - {@link Errors#INVALID_REQUEST} if the last epoch or offset are invalid
```
My reasoning is that this error is reserved for cases that are not expected to happen with a proper functioning application. So whenever we see `INVALID_REQUEST`, it indicates a bug. The case of `INCONSISTENT_VOTER_SET` is a bit different because we could definitely see this in the middle of a reassignment (once we've added supported).

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -411,52 +504,85 @@ private EndQuorumEpochResponseData handleEndQuorumEpochRequest(EndQuorumEpochReq
     }
 
-
+    /**
```
Sure, I can add something to the comment. Note that one change I made here is that we no longer return the out of range error. Instead, we expect the follower to always check the next offset field. I felt this was a more consistent way of handling this case.

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -329,8 +349,15 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
-                voteGranted = lastEpochEndOffset.compareTo(endOffset()) >= 0;
+                int lastEpoch = request.lastEpoch();
+                long lastEpochEndOffset = request.lastEpochEndOffset();
+                if (lastEpochEndOffset < 0 || lastEpoch < 0 || lastEpoch >= request.candidateEpoch()) {
```
I was considering this. The checks in `handleFetchQuorumRecords` and `handleVoteRequest` are slightly different. We can be stricter in `handleVoteRequest` because a candidate cannot have any appended entries from the epoch of its candidacy.

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -339,26 +366,48 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
-        if (quorum.isLeader()) {
-            logger.debug("Ignoring vote response {} since we already became leader for epoch {}",
+    /**
+     * Handle a received Vote response.
```
True, this is mostly filler since I decided to add the javadoc. Since these methods are all following the same convention, perhaps there is no need to document each one.

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -292,12 +292,20 @@ private void onBecomeFollowerOfElectedLeader(FollowerState state) {
         }
     }
 
+    private void becomeFollower(OptionalInt leaderId, int epoch) throws IOException {
```
I would say no harm in keeping it? It does server to make the logic more concise in the caller.

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -339,26 +366,42 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
+    private boolean handleVoteResponse(
+        RaftResponse.Inbound responseMetadata
+    ) throws IOException {
+        int remoteNodeId = responseMetadata.sourceId();
```
As we discussed earlier today, it makes sense to consider introducing some inheritance property into the auto generated protocol for writing less boiler-template code.

hachikuji replied:
Yes. This would be helpful in the case of the error code. But it's probably not a blocking feature gap.

abbccdda commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -532,45 +655,59 @@ private OptionalLong maybeAppendAsLeader(LeaderState state, Records records) {
+        int responseEpoch = response.leaderEpoch();
+
+        // Always update voter connections if they are present, regardless of errors
+        updateVoterConnections(response);
```
I seem to overlook this case, why does it make sense to always update connection information? Should we always look for response with equal or greater epoch?

A more general question is,  we do expect a broker to change their connection information through redeployment, which means the validity of connection information also goes down over time. Maybe we don't need to be strict in V1, but I feel in the long run only accepting up-to-date connection information seems less error-prone.

hachikuji replied:
The `bootTimestamp` is what we use to determine if a node's connection information is newer, so it's safe to use without considering epoch. The simulation tests fail without this check because we get stuck in states where one node that can't connect to other nodes and starts holding elections. However, the bumped epoch from these elections would prevent it from getting the latest information. Hence we update connections always.

I'm not super thrilled with the `bootTimestamp` proposal for what it's worth. I would prefer to have something which could guarantee monotonicity, but we also need something which can be derived from only local state. The best idea I've thought of is to store a separate epoch which is bumped on every boot. But that also is not ideal.

abbccdda replied:
I guess boot timestamp is good for now, although from our offline discussion it is not 100% reliable, but we could delay the solution until we discover more real issue under system tests.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -375,21 +418,48 @@ private BeginQuorumEpochResponseData buildBeginQuorumEpochResponse(Errors error)
+        RaftRequest.Inbound requestMetadata
+    ) throws IOException {
+        BeginQuorumEpochRequestData request = (BeginQuorumEpochRequestData) requestMetadata.data;
+        Optional<Errors> errorOpt = validateVoterOnlyRequest(request.leaderId(), request.leaderEpoch());
```
I actually agrees with your previous TODO marker inside this function to not call `becomeUnattachedFollower` inside, and one reason I had is that here, we will call `becomeUnattachedFollower` first and then `becomeFetchingFollower` immediately after.

What about this (admittedly it is also a bit awkward): let this function returns `UNKNOWN_LEADER_EPOCH` if `requestEpoch > quorum.epoch()`, and the caller can decide whether to really encode this error or not --- e.g. in this caller we would just ignore it and call `becomeFetchingFollower` afterwards, whereas in the `handleVoteRequest` caller above we would then call `becomeUnattachedFollower` in the caller.

hachikuji replied:
Let me see if some name changes can make the `becomeUnattachedFollower` transition less surprising. Note that semantically there is no problem with a call to `becomeUnattachedFollower` followed by `becomeFetchingFollower`.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -306,10 +314,22 @@ private VoteResponseData buildVoteResponse(Errors error, boolean voteGranted) {
+        RaftRequest.Inbound requestMetadata
+    ) throws IOException {
+        VoteRequestData request = (VoteRequestData) requestMetadata.data;
+        Optional<Errors> errorOpt = validateVoterOnlyRequest(request.candidateId(), request.candidateEpoch());
```
Related to my other comment: I feel it is a bit awkward that `validateVoterOnlyRequest` may change this voter from a leader state to an unattached follower, and then in line 335 below the check `quorum.isLeader()` would be void --- readers may expect that the `validateXXX` function would not change the state of this voter.

I'd suggest we do not call `becomeUnattachedFollower` inside the validate function, and then let the this function, the caller to decide based on its current state whether the returned error code can be handled accordingly or should be returned in the error response.

hachikuji replied:
This response is also related to my other response 🙂 . The invariant that `validateVoterOnlyRequest` and `validateLeaderOnlyRequest` provide is that if no error is returned, then the request is guaranteed to match the current epoch. I agree the state change is a little surprising given the name of the function, but maybe that just means we need a better name. The alternative is to give up the invariant and duplicate the checks. I will think about it a little more, but I'm not too eager to give this up because I felt it really did reasoning about the states simpler.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -306,10 +314,22 @@ private VoteResponseData buildVoteResponse(Errors error, boolean voteGranted) {
+        Optional<Errors> errorOpt = validateVoterOnlyRequest(request.candidateId(), request.candidateEpoch());
         if (errorOpt.isPresent()) {
-            return buildVoteResponse(Errors.forException(errorOpt.get()), false);
+            return buildVoteResponse(errorOpt.get(), false);
```
This is for line 336/340 below: I think these lines would only happen if the maintained epoch is exactly the same as the request epoch (if it is larger we'd return `FENCED_LEADER_EPOCH` in line 332, if it is smaller we'd no longer be leader / candidate any more). If that's the case then we could make the log4j to be more explicit like "... of epoch {} which is the same as the request epoch".

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +721,244 @@ private void updateVoterConnections(FindQuorumResponseData response) {
+     *    - Optional.of(false) indicates that the response was handled here, but that the request
+     *        will need to be retried
+     */
+    private Optional<Boolean> maybeHandleStandardError(
```
nit: my paranoid hat is urging me to suggest rename it to `maybeHandleCommonResponse` to be aligned with `validateLeaderOnlyRequest / validateVoterOnlyRequest`.

Also, at the end could we check that if the `error` passed in is NONE then we can just return `Optional.of(true)` and hence the caller do not need to check this anymore?

guozhangwang replied:
More generally speaking, I think the contract should be `if there's any issues indicated in the other fields, then the error field should always not be NONE, though vice versa may not be true`, and the error handling logic can be simpler.

hachikuji replied:
Ack on the naming. I am not sure I follow your other point though. Having no error is a necessary, but insufficient condition to trigger the rest of the response handling logic. The intended semantic is `Optional.of(value)` means we handled it here and we don't need to bother with any additional handling. On the other hand, `Optional.empty()` means that all basic validations passed and we need to handle the case with the respective API semantics in mind.

I realize this may fall short of perfection, but I think it is a reasonable compromise to ensure consistent handling of common cases.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -339,26 +366,42 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
             }
+            return true;
+        } else {
+            return handleUnexpectedError(error, responseMetadata);
```
If we can guarantee that if there's any fields in the response that indicate an error that should be handled, then the `error` field is never NONE, then we would never hit this path. Ditto elsewhere.

guozhangwang replied:
Generally speaking, the 

```
        Optional<Boolean> handled = maybeHandleStandardError(..)
        if (handled.isPresent()) {
            return handled.get();
        } else if (error == Errors.NONE) {
            ...
            return true;
        } else {
            return handleUnexpectedError(error, responseMetadata);
        }
```

pattern can be simplified as

```
if (error == NONE) {
    // ... extra logic
    return true;
} else {
    handled = maybeHandleStandardError(..)  // the error passed in is guaranteed to be not NONE.
    if (handled.isPresent()) {
        return handled.get()
    } else {
        //.. specific handling by the caller, potentially calling handleUnexpectedError
    }
}
```

Personally I feel the above is more straight-forward to understand.

hachikuji replied:
I understand where you're coming from, but one of the annoying things is the need to check the epoch in all cases. Even if the error is NONE, we may no longer be interested in the response because we have changed states. I think the nice guarantee that `maybeHandleStandardError` gives you is that the epoch of the response matches the epoch from the request. I found that made it easier to reason about the cases that needed to be handled.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +721,244 @@ private void updateVoterConnections(FindQuorumResponseData response) {
-            // Nothing to do for now
-        } else if (responseData instanceof FindQuorumResponseData) {
-            handleFindQuorumResponse((FindQuorumResponseData) responseData);
+        ApiKeys apiKey = ApiKeys.forId(response.data.apiKey());
```
nit: add log.trace recording the recipient of the response?

hachikuji replied:
There is a trace message under `sendOutboundMessage`.

guozhangwang commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -584,211 +721,244 @@ private void updateVoterConnections(FindQuorumResponseData response) {
 
     private void handleRequest(RaftRequest.Inbound request) throws IOException {
-        ApiMessage requestData = request.data();
+        ApiKeys apiKey = ApiKeys.forId(request.data.apiKey());
```
nit: log4j.trace recording the recipient of the request?

hachikuji replied:
I have a trace message under `handleInboundMessage`.
